### PR TITLE
Stack exchange provider

### LINKED ISSUE #16: Add StackExchange provider

### PR DESCRIPTION
**Pull request by andrewlock**
Fixes #16 

Some bits of this are a little scrappy and may require some additional tidying up, but it works and is tested 😄 

There were basically 2 pain points - the StackExchange TokenEndpoint ignores the `accept` header, and always sends back a string of the form `access_token=...&expires=1234`, so that needs to be handled in `ExchangeCodeAsync`.

Also, the response from the UserInformationEndpoint is both gzipped and utf-8 encoded, which is a bit of a pain, and requires a reference to System.IO.Compression. The handler at the moment assumes that the response will always be utf-8 and gzipped, but we may want to improve it to handle future changes if the api stops doing that?

Finally, there's not actually that many useful claims you can get back - you get an id and a display name, but you can't get FirstName, LastName or Email. 

Still, it works at least!

### CODE REVIEW

**Comment by andrewlock** [comment]
It seems the `alpha` mono build is stalling, though the `latest` build completes successfully. There's no errors or anything, it just seems to stop running. I've seen this before with the `alpha` mono release but I never managed to get to the bottom of it.. any ideas?

**Comment by PinpointTownes** [comment]
Thanks for your contribution!

Looks like StackExchange is going to win both the "most stupid API ever" and the "non-standard OAuth2 implementation" awards this year :trollface: 

> Also, the response from the UserInformationEndpoint is both gzipped and utf-8 encoded, which is a bit of a pain, and requires a reference to System.IO.Compression. The handler at the moment assumes that the response will always be utf-8 and gzipped, but we may want to improve it to handle future changes if the api stops doing that?

Indeed, that seems a bit fragile (:gun: the SO guys for not respecting the HTTP spec...).

That said, isn't `HttpClient` supposed to automatically decompress the HTTP response? I thought both GZIP and DEFLATE were enabled by default (well, at least in CoreFX: https://github.com/dotnet/corefx/blob/master/src/Common/src/System/Net/Http/HttpHandlerDefaults.cs#L15).

> Finally, there's not actually that many useful claims you can get back - you get an id and a display name, but you can't get FirstName, LastName or Email.

Doesn't SE have additional scopes we could use for that?

**Comment by PinpointTownes** [comment]
> I've seen this before with the alpha mono release but I never managed to get to the bottom of it.. any ideas?

Nope, no idea. This issue is impacting all our repos and I have no idea why...

**Comment by andrewlock** [comment]
> Looks like StackExchange is going to win both the "most stupid API ever" and the "non-standard OAuth2 implementation" awards this year

Haha, yep it turned into bit more of a trial than I anticipated! 

> That said, isn't `HttpClient` supposed to automatically decompress the HTTP response

Hmmm, yeah that's a good point - I'll dig more into that one, I think I can probably do something with it based on the comments here: https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/HttpContent.cs#L179

> Doesn't SE have additional scopes we could use for that?

Unfortunately not, apparently they use to expose them but they removed them due to privacy concerns. 

According to the docs http://api.stackexchange.com/docs/authentication, the additional scopes allow access to things like their inbox and write access etc. 

> Nope, no idea. This issue is impacting all our repos and I have no idea why...

Gah, how frustrating:s

File: src/AspNet.Security.OAuth.StackExchange/StackExchangeAuthenticationOptions.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            AuthenticationScheme = StackExchangeAuthenticationDefaults.AuthenticationScheme;
+            DisplayName = StackExchangeAuthenticationDefaults.DisplayName;
+            ClaimsIssuer = StackExchangeAuthenticationDefaults.Issuer;
+
+            CallbackPath = new PathString(StackExchangeAuthenticationDefaults.CallbackPath);
+
+            AuthorizationEndpoint = StackExchangeAuthenticationDefaults.AuthorizationEndpoint;
+            TokenEndpoint = StackExchangeAuthenticationDefaults.TokenEndpoint;
+            UserInformationEndpoint = StackExchangeAuthenticationDefaults.UserInformationEndpoint;
+            BackchannelHttpHandler = new HttpClientHandler {AutomaticDecompression = DecompressionMethods.GZip};
```
Comment (andrewlock) [inline src/AspNet.Security.OAuth.StackExchange/StackExchangeAuthenticationOptions.cs]: This line lets me get around having to do the gzip decompression and encoding manually.

**Comment by andrewlock** [comment]
@PinpointTownes with the exception of the build failure, I think this should be good to merge now?

**Comment by PinpointTownes** [comment]
Merged: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/commit/a890142fb068e0129d628f4a8d31891a238c6e4b

Thanks for your contribution :clap:

**Pull request closed without merging**